lib/typing/mongo/bson.go

```diff
@@ -32,133 +27,101 @@ func JSONEToMap(val []byte) (map[string]any, error) {
+		result[i] = value
 	}
-
-	return vw.WriteString(s.Hex())
+	return result, nil
 }
 
-func binaryEncodeValue(_ bsoncodec.EncodeContext, vw bsonrw.ValueWriter, val reflect.Value) error {
-	if !val.IsValid() || val.Type() != tBinary {
-		return bsoncodec.ValueEncoderError{Name: "ObjectIDEncodeValue", Types: []reflect.Type{tBinary}, Received: val}
-	}
-
-	s, isOk := val.Interface().(primitive.Binary)
-	if !isOk {
-		return bsoncodec.ValueEncoderError{Name: "ObjectIDEncodeValue not Binary", Types: []reflect.Type{tBinary}, Received: val}
-	}
-
-	switch s.Subtype {
+func bsonBinaryValueToMap(value primitive.Binary) (any, error) {
+	switch value.Subtype {
 	case
 		bson.TypeBinaryUUIDOld,
 		bson.TypeBinaryUUID:
-		parsedUUID, err := uuid.FromBytes(s.Data)
+		parsedUUID, err := uuid.FromBytes(value.Data)
 		if err != nil {
-			return err
+			return nil, err
 		}
-		return vw.WriteString(parsedUUID.String())
+
+		return parsedUUID.String(), nil
 	default:
-		return vw.WriteBinaryWithSubtype(s.Data, s.Subtype)
+		return map[string]any{
+			"$binary": map[string]any{
+				"base64":  base64.StdEncoding.EncodeToString(value.Data),
+				"subType": fmt.Sprintf("%02x", value.Subtype),
+			},
+		}, nil
 	}
 }
 
-func timestampEncodeValue(_ bsoncodec.EncodeContext, vw bsonrw.ValueWriter, val reflect.Value) error {
-	if !val.IsValid() || val.Type() != tTimestamp {
-		return bsoncodec.ValueEncoderError{Name: "ObjectIDEncodeValue", Types: []reflect.Type{tBinary}, Received: val}
-	}
-
-	s, isOk := val.Interface().(primitive.Timestamp)
-	if !isOk {
-		return bsoncodec.ValueEncoderError{Name: "ObjectIDEncodeValue not Binary", Types: []reflect.Type{tBinary}, Received: val}
-	}
+func bsonValueToGoValue(value any) (any, error) {
+	switch v := value.(type) {
+	case primitive.DateTime:
+		return time.Unix(0, int64(v)*int64(time.Millisecond)).UTC().Format(ext.ISO8601), nil
+	case primitive.ObjectID:
+		return v.Hex(), nil
+	case primitive.Binary:
+		return bsonBinaryValueToMap(v)
+	case primitive.Decimal128:
+		potentialFloat, err := strconv.ParseFloat(v.String(), 64)
+		if err != nil {
+			return nil, err
+		}
 
-	return vw.WriteString(time.Unix(int64(s.T), 0).UTC().Format(ext.ISO8601))
-}
+		if v.String() == fmt.Sprint(potentialFloat) {
+			return potentialFloat, nil
+		}
 
-func createCustomRegistry() *bsoncodec.Registry {
-	rb := bson.NewRegistry()
-	rb.RegisterTypeEncoder(tDateTime, bsoncodec.ValueEncoderFunc(dateTimeEncodeValue))
-	rb.RegisterTypeEncoder(tOID, bsoncodec.ValueEncoderFunc(objectIDEncodeValue))
-	rb.RegisterTypeEncoder(tBinary, bsoncodec.ValueEncoderFunc(binaryEncodeValue))
-	rb.RegisterTypeEncoder(tDecimal, bsoncodec.ValueEncoderFunc(decimalEncodeValue))
-	rb.RegisterTypeEncoder(tTimestamp, bsoncodec.ValueEncoderFunc(timestampEncodeValue))
-	return rb
+		// If the string representation of the Decimal128 is not equal to the float representation
```

Comment:
In what cases would the string representation be different? Is this something we expect to happen often or something we should log if it happens?

Reply:
On a practical level, float64 should be sufficient. However, in theory a decimal128 has more precision than float64.

So it is possible that a number becomes looses precision with f64

Reply:
https://en.wikipedia.org/wiki/Decimal128_floating-point_format 

https://en.wikipedia.org/wiki/Decimal64_floating-point_format

Reply:
I see, is there a disadvantage to just always returning a string then? I feel like it might be harder to work with these values if sometimes they are float64s and sometimes they are strings.

Another thing I'm wondering is whether we'll end up with false positives for relatively small numbers such as `1234567.0` since when they are run through `fmt.Sprintf()` they will be output in exponential notation (e.g. `1.234567e+06`) and then will be different from what `Decimal128.String()` outputs.

Reply:
I'm not opposed to the idea of outputting strings to be more deterministic. BSON also has `int64/float64` that folks can use

Reply:
Convert to string then?